File: backend/app/api/routes/weekly_plans.py
"""Weekly learning plan routes."""

from typing import Union
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.ai.openai_provider import OpenAIProvider
from app.api.deps import get_current_user_id
from app.core.database import get_db
from app.schemas.weekly_plan import RoadmapCompletionSummary, WeeklyPlan
from app.services.weekly_plan_service import WeeklyPlanService

router = APIRouter()


# ── Dependency helpers ─────────────────────────────────────────────────────────


def get_weekly_plan_service(
    db: AsyncSession = Depends(get_db),
) -> WeeklyPlanService:
    """Dependency that constructs a WeeklyPlanService with the current DB session and AI provider."""
    ai_provider = OpenAIProvider()
    return WeeklyPlanService(db=db, ai_provider=ai_provider)


# ── Endpoints ──────────────────────────────────────────────────────────────────


@router.get(
    "",
    response_model=list[WeeklyPlan],
    summary="List all weekly plans",
)
async def list_weekly_plans(
    user_id: UUID = Depends(get_current_user_id),
    service: WeeklyPlanService = Depends(get_weekly_plan_service),
) -> list[WeeklyPlan]:
    """Retrieve all weekly plans for the authenticated user's current roadmap.

    Returns plans ordered by week number.

    Domain exceptions (NoGapAnalysisError) are handled by the registered
    exception handlers automatically.

    Requirements: 5.1, 5.2, 5.3, 5.4, 5.5, 5.6
    """
    return await service.get_all_plans(user_id=user_id)


@router.get(
    "/current",
    response_model=WeeklyPlan,
    summary="Get the current active weekly plan",
)
async def get_current_plan(
    user_id: UUID = Depends(get_current_user_id),
    service: WeeklyPlanService = Depends(get_weekly_plan_service),
) -> WeeklyPlan:
    """Retrieve the user's active (in-progress) weekly plan.

    Returns 404 if there is no active plan.

    Requirements: 5.1, 5.2, 5.3
    """
    plan = await service.get_current_plan(user_id=user_id)
    if plan is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={
                "error": "NO_ACTIVE_PLAN",
                "message": "No active weekly plan found.",
            },
        )
    return plan


@router.put(
    "/{plan_id}/tasks/{task_id}/complete",
    response_model=Union[WeeklyPlan, RoadmapCompletionSummary],
    summary="Mark a task as complete",
)
async def complete_task(
    plan_id: UUID,
    task_id: UUID,
    user_id: UUID = Depends(get_current_user_id),
    service: WeeklyPlanService = Depends(get_weekly_plan_service),
) -> Union[WeeklyPlan, RoadmapCompletionSummary]:
    """Mark a specific task within a weekly plan as completed.

    When all tasks in a plan are completed, the plan status advances to
    'completed' and the next plan becomes 'in-progress'. If this was the
    final plan in the roadmap, a RoadmapCompletionSummary is returned instead.

    Domain exceptions (WeeklyPlanNotFoundError, WeeklyTaskNotFoundError) are
    handled by the registered exception handlers automatically.

    Requirements: 5.3, 5.4, 5.6
    """
    return await service.mark_task_complete(plan_id=plan_id, task_id=task_id, user_id=user_id)


@router.post(
    "/adjust",
    response_model=list[WeeklyPlan],
    summary="Adjust remaining plans for delay",
)
async def adjust_plan(
    user_id: UUID = Depends(get_current_user_id),
    service: WeeklyPlanService = Depends(get_weekly_plan_service),
) -> list[WeeklyPlan]:
    """Adjust remaining weekly plans to accommodate incomplete tasks from the current plan.

    Redistributes incomplete tasks from the current in-progress plan into
    upcoming plans. Does not require a request body since it operates on the
    user's current in-progress plan.

    Domain exceptions (NoGapAnalysisError) are handled by the registered
    exception handlers automatically.

    Requirements: 5.5
    """
    return await service.adjust_for_delay(user_id=user_id)

File: frontend/src/lib/api/weekly-plans.ts
import type { WeeklyPlan, AdjustDelayResult } from "@/types/weekly-plan";
import { jsonAuthHeaders } from "./client";

const API_BASE = "/api/weekly-plans";

export class WeeklyPlansApiError extends Error {
  public code: string;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    this.name = "WeeklyPlansApiError";
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const body = await response.json().catch(() => null);
    const detail = body?.detail ?? body;
    const code = detail?.error ?? "UNKNOWN";
    const message = detail?.message ?? "An unexpected error occurred";
    throw new WeeklyPlansApiError(code, message);
  }
  return response.json();
}

/**
 * Get all weekly plans for the current roadmap.
 * GET /api/weekly-plans
 */
export async function getWeeklyPlans(): Promise<WeeklyPlan[]> {
  const response = await fetch(API_BASE, {
    method: "GET",
    headers: jsonAuthHeaders(),
  });
  return handleResponse<WeeklyPlan[]>(response);
}

/**
 * Get the current active weekly plan.
 * GET /api/weekly-plans/current
 */
export async function getCurrentWeeklyPlan(): Promise<WeeklyPlan> {
  const response = await fetch(`${API_BASE}/current`, {
    method: "GET",
    headers: jsonAuthHeaders(),
  });
  return handleResponse<WeeklyPlan>(response);
}

/**
 * Mark a task as complete.
 * PUT /api/weekly-plans/{planId}/tasks/{taskId}/complete
 */
export async function markTaskComplete(
  planId: string,
  taskId: string
): Promise<WeeklyPlan> {
  const response = await fetch(`${API_BASE}/${planId}/tasks/${taskId}/complete`, {
    method: "PUT",
    headers: jsonAuthHeaders(),
  });
  return handleResponse<WeeklyPlan>(response);
}

/**
 * Adjust remaining plans when there's a delay.
 * POST /api/weekly-plans/adjust
 */
export async function adjustForDelay(): Promise<AdjustDelayResult> {
  const response = await fetch(`${API_BASE}/adjust`, {
    method: "POST",
    headers: jsonAuthHeaders(),
  });
  return handleResponse<AdjustDelayResult>(response);
}
